src/NUnitFramework/framework/Attributes/XmlDataSourceAttribute.cs

```diff
@@ -0,0 +1,104 @@
+﻿// ***********************************************************************
+// Copyright (c) 2008-2015 Charlie Poole
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Xml;
+using System.Collections;
+using System.Reflection;
+using NUnit.Framework.Interfaces;
+using NUnit.Framework.Internal;
+using NUnit.Framework.Internal.Builders;
+
+
+namespace NUnit.Framework.Attributes
+{
+    /// <summary>
+    /// Import data from XML document
+    /// </summary>
+    public class XmlDataAttribute : TestCaseSourceAttribute
+    {
+        private XmlData xml;
+
+        /// <summary>
+        /// Number of rows to read
+        /// </summary>
+        public int RowsToRead
+        {
+            get { return xml.RowsToRead; }
+            set { xml.RowsToRead = value; }
+        }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="source"></param>
+        public XmlDataAttribute(string source) : base(source)
+        {
+            xml = new XmlData(source);
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="methodInfo"></param>
+        /// <returns></returns>
+        protected override IEnumerable GetTestCaseSource(IMethodInfo methodInfo)
+        {
+            if (!xml.SourceFileExists)
+            {
+                string errMsg = String.Format("Unable to find source: {0}", xml.Name);
+                throw new FileNotFoundException(errMsg);
+            }
+
+            // get dimensions from method parameters
+            var paramInfo = methodInfo.GetParameters();
+            var paramDim = paramInfo.Length;
+
+
+            var dataFromQuery = xml.GetData();
+
+            // check if items were returned
+            if (((IList)dataFromQuery).Count == 0)
```

Comment:
I don't like that you have internal knowledge that the returned data is an IList. Why not just switch the return type to be IList?